### ISSUE TITLE
Nested conditional types and generics, testing type equality evaluating to incorrect branch

### ISSUE THREAD

**Issue by AnyhowStep** [microsoft/TypeScript#37314]
**TypeScript Version:**  3.5.1, 3.8.3

**Search Terms:**
conditional types, generics, equality

**Code**

strictNullChecks and strictFunctionTypes enabled
```ts
export type Equals<A1 extends any, A2 extends any> =
    (<A>() => A extends A1 ? 1 : 0) extends (<A>() => A extends A2 ? 1 : 0)
    ? 1
    : 0

interface Foo<T> {
    x : () => T
}

declare const a: Foo<Date>;
declare const b: Foo<Date | null>;

//Expected 0, Actual 1
type ShouldBe0 = Equals<typeof a, typeof b>;
```

**Expected behavior:**

Evaluates to `0`

**Actual behavior:**

Evaluates to `1`

**Playground Link:** http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=16&pc=1#code/FAUwHgDg9gTgLgAjgTwiBBRAjgVwIYA2AzgDwCCAjAuHCAHYAmRCedyANAmQEzVi2NmrZAD4EAXmAJpCABTkRsgJQSxZPgKZcqAfgRUAXAgAMKmvS3yyileLUaLzHgj2GTSqTNefpR48GAASzpaGAAzPABjdAAxKCgSABUxAG8fBDAEI2VVBETgAF8AhhBIgjwYdEioOiJEPCM4hIARPFoRAG5gErKKqpq6hAAjRviSVtoEAB8EOhwCAk6AgHpljEhS2gYTTjJIuHwCfWAUNAQAZQALKHmGACEQYwlMXEJSU5AoMJZOD6-hpZAA

**Related Issues:** Not that I know of

I filed this issue over here as well, https://github.com/pirix-gh/ts-toolbelt/issues/97

-----

Looking at the type, it **should** work. But it doesn't. I think it might indicate something is broken with TS

**Comment by AnyhowStep** [comment]
Type from here, https://github.com/Microsoft/TypeScript/issues/27024#issuecomment-421529650

So, https://github.com/Microsoft/TypeScript/issues/27024 is related

**Comment by AnyhowStep** [comment]
The comment was made on September 2018.

Around that time, TS 3.0 and TS 3.1 were out.

It also breaks for TS 3.0 and TS 3.1

[TS 3.0 Playground](http://www.typescriptlang.org/play/?ts=3.0.1&ssl=1&ssc=1&pln=22&pc=1#code/FAUwHgDg9gTgLgAjgTwiBBRAjgVwIYA2AzgDwCCAjAuHCAHYAmRCedyANAmQEzVi2NmrZAD4EAXmAJpCABTkRsgJQSxZPgKZcqAfgRUAXAgAMKmvS3yyileLUaLzHgj2GTSqTNefpR48GAASzpaGAAzPABjdAAxKCgSABUxAG8fBDAEI2VVBETgAF8AhhBIgjwYdEioOiJEPCM4hIARPFoRAG5gErKKqpq6hAAjRviSVtoEAB8EOhwCAk6AgHpljEhS2gYTTjJIuHwCfQCTgJQ0BABlAAsoeYYAIRBjCUxcQlJzkCgwlk4vn7DJZAA)

[TS 3.1 Playground](http://www.typescriptlang.org/play/?ts=3.1.6&ssl=1&ssc=1&pln=22&pc=1#code/FAUwHgDg9gTgLgAjgTwiBBRAjgVwIYA2AzgDwCCAjAuHCAHYAmRCedyANAmQEzVi2NmrZAD4EAXmAJpCABTkRsgJQSxZPgKZcqAfgRUAXAgAMKmvS3yyileLUaLzHgj2GTSqTNefpR48GAASzpaGAAzPABjdAAxKCgSABUxAG8fBDAEI2VVBETgAF8AhhBIgjwYdEioOiJEPCM4hIARPFoRAG5gErKKqpq6hAAjRviSVtoEAB8EOhwCAk6AgHpljEhS2gYTTjJIuHwCfQCTgJQ0BABlAAsoeYYAIRBjCUxcQlJzkCgwlk4vn7DJZAA)

**Comment by weswigham** [comment]
Hm, so this is fixed in `master`, or at least it no longer reproduces. @RyanCavanaugh do we want to track down and backport the fix to 3.8?

**Comment by AnyhowStep** [comment]
I'm wondering if this should be added to the test suite, just to be safe =x
